File: Supervised-Learning/white_wine_logistic_regression.py
# Vanilla Logistic Regression with Train-Test Split and Cross-Validation
# ----------------------------------------------------------------------

# Import Packages
# ---------------

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_score

# Import and Explore Data
# ---------------------------

df = pd.read_csv("/Users/Timo/Documents/Propulsion_Academy/Teaching/Batch_3/dswd-2018-04/05-Machine-Learning-2/00_Data/winequality-white.csv", sep=";")

print(df.info())

# Check Response Variable Distribution
# ------------------------------------

df["quality"].hist()

# Verdict: Classes are highly unbalanced
# Solution: Undersampling of minority classes since observations < 10000

# Check distributions of predictor variables
# ------------------------------------------

plt.figure()
plt.suptitle("Distributions of Values for each column")
for i in range(1, len(df.columns) - 1):
    plt.subplot(10, 1, i)
    ax = sns.distplot(df[df.columns[i]], kde=False, rug=False)
    ax.set_title(df.columns[i])
    ax.set_xlabel("")
plt.show()

# Verdict: The "chlorides" variable is unbalanced.
# Solution: Correct with log-transformation

df["chlorides"] = np.log10(df["chlorides"])
df["chlorides"].hist()

# Perform Cross Validation with Vanilla Logistic Regression
# ---------------------------------------------------------

X = df.drop("quality", axis=1)
y = df["quality"]

sm = SMOTE(random_state=42, k_neighbors=4)

X_over, y_over = sm.fit_sample(X.values, y.values)

scaler = StandardScaler()
scaler.fit(X_over)

X_scaled = scaler.transform(X_over)

model = LogisticRegression()

scores = cross_val_score(model, X_scaled, y_over, cv=10)
accuracy = np.mean(scores)

print(accuracy)

# Penalize Model
# --------------

l1_model = LogisticRegression(C=1.0, penalty='l1', tol=1e-6)
l1_scores = cross_val_score(l1_model, X_scaled, y_over, cv=10)
l1_accuracy = np.mean(l1_scores)
print(l1_accuracy)

l2_model = LogisticRegression(penalty="l2")
l2_scores = cross_val_score(l2_model, X_scaled, y_over, cv=10)
l2_accuracy = np.mean(l2_scores)
print(l2_accuracy)

File: Supervised-Learning/white_wine_logistic_regression.R
# Vanilla Logistic Regression with Train-Test Split and Cross-Validation
# ----------------------------------------------------------------------

setwd("~/Documents/Education/Machine-Learning-Projects/Supervised-Learning/")
set.seed(42)

# Load Libraries
# --------------

library(tidyverse)
library(ggthemes)
library(gridExtra)
library(caret)
library(nnet)

# Import and Explore Data
# -----------------------

df <- read.csv("winequality-white.csv", sep = ";")
str(df)

# Check Response Variable Distribution
# ------------------------------------

table(df$quality)
df$quality <- as.factor(df$quality)

# Verdict: Classes are highly unbalanced
# Solution: Undersampling of minority classes since observations < 10000

# Check distributions of predictor variables
# ------------------------------------------

plotter <- function(df, col) {
  
  title <- names(df)[col]
  
  p <- ggplot(df, aes(x = df[, col])) + 
    geom_histogram(stat="count") +
    ggtitle(title) +
    xlab("") +
    ylab("Count")
    theme_tufte()
  
  return(p)
}

plots <- lapply(seq(1:ncol(df)), function(x)(plotter(df, x)))
grid.arrange(grobs = plots, ncol = 2)

# Verdict: The "chlorides" variable is unbalanced.
# Solution: Correct with log-transformation

df$chlorides <- log10(df$chlorides)

plotter(df, 5)

# Normalize Data
# --------------

y <- df$quality
X <- select(df, -quality)
df_scaled <- data.frame(scale(X), y)

# Perform Cross Validation with Vanilla Logistic Regression
# ---------------------------------------------------------

idx <- createDataPartition(df_scaled$y, p = 0.7, list = FALSE)
training_data <- df_scaled[idx,]
testing_data <- df_scaled[-idx,]

cross_validation <- trainControl(method = "repeatedcv", # Cross-validation method
                                 number = 10, # Number of k-folds
                                 repeats = 10, # Number of repetition
                                 verboseIter = FALSE, # Display text
                                 sampling = "smote" # Balance classes
) 

model.fit <- train(y ~ ., # Specify model
                   data = training_data, # Use training data
                   method = "multinom", # Model: Multinomial logistic regression (nnet package)
                   trControl = cross_validation # Use pre-defined training control from above
) 

model.predict <- predict(model.fit, # Use trained model
                         newdata = testing_data # Pass test data to function
                         )

model.fit